### PR TITLE
feat: Add custom_fetch to allow alternate WebAPI Fetch implementation

### PR DESCRIPTION
**Pull request by agrohs**
## Type

* ### Feature
  Implements a new feature

## Description
_A brief description of the goals of the pull request._
* Allows `custom_fetch` to be passed into `MoltinGateway` configuration that will be used for authentication calls to API. This allows alternate implementations of the WebAPI Fetch (https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) spec to be used.

```
    const Moltin = MoltinGateway({
      client_id: 'X',
      client_secret: 'X',
      grant_type: 'client_credentials',
      custom_fetch: myCustomWebAPIfetch
    })
```

## Dependencies
_Other PRs or builds that this PR depends on._
* N/A

## Issues

*A list of issues closed by this PR.*

* N/A

## Notes

* Ex. - a scenario where you want to proxy the request for a `client_credentials` token to avoid exposing the `client_secret` in any requests.

### CODE REVIEW

**Comment by agrohs** [comment]
@gje4 - just an FYI, this is the PR I was mentioning to support the secure WebAPI Fetch implementation

**Comment by ynnoj** [comment]
@agrohs Thanks for the pull request. 

I’m struggling to understand a scenario when passing in another fetch interface would be required. Can you describe the use case in your situation, perhaps provide some implementation examples?

Thanks!

**Comment by agrohs** [comment]
No worries, I figured it might come out of context. In the example in the Notes at the bottom of the PR, described the scenario that we are looking at (though I could imagine a case where others might have valid reasons to have their own custom implementation). In our case, we are simply sending requests to a proxy that has access to private keys and contains a token replacement system it uses before they are sent to the original requested URL (with all headers and such intact). This ensure the client secret is not exposed on the client side, but are replaced before the proxied fetch from there goes through to the original Moltin API call. Happy to hop on a call to talk through and show you the custom implementation we are sending in/why we need it?

George has a bit more background before this was submitted and all has been tested on our end using the custom Fetch implementation to route through the proxy.


ADAM GROHS.
CO-FOUNDER/CEO
646.244.4501

Particular.
the innovation agency of tomorrow, today

www.uniquelyparticular.com

**Comment by ynnoj** [comment]
> In our case, we are simply sending requests to a proxy that has access to private keys and contains a token replacement system it uses before they are sent to the original requested URL (with all headers and such intact)

If there’s already some server element involved, why not just make the requests that require secret keys from there and query that server from the client?

**Comment by agrohs** [comment]
We don’t have any sever element on our end, but a product it’s being wrapped in that provides a proxy method to secure keys for this exact purpose. We’ve simply wrapped their proxy in an implementation of the Fetch spec to make it easier to plug in here (and elsewhere).

www.uniquelyparticular.com

**Comment by ynnoj** [comment]
@agrohs What does the `myCustomWebAPIfetch` in your example look like? Which fields do you need to override that are implemented by the standard fetch interface in the auth request?

File: src/factories/request.js [hunk 1]
```diff
@@ -70,6 +71,7 @@ class RequestFactory {
         config.client_id,
         response.access_token,
         response.expires
+        // Math.min(response.expires, config.auth.expires)
```
Comment (notrab) [inline src/factories/request.js]: @agrohs Is this required?

File: src/config.js [hunk 2]
```diff
@@ -9,7 +9,8 @@ class Config {
       client_secret,
       currency,
       host,
-      storage
+      storage,
+      secure_fetch
```
Comment (notrab) [inline src/config.js]: @agrohs @ynnoj is `secure_fetch` the best name for this param? Would `custom_fetch` work?

**Comment by notrab** [comment]
@agrohs Are you able to provide a basic outline of `myCustomWebAPIfetch`? I can use that in writing a test and get this merged.

**Comment by agrohs** [comment]
Yep, I’m not partial as to what it’s called (actually had as something like custom_fetch earlier, but as it is used for the auto call was playing with auth_fetch or secure_fetch or credentials_fetch).

ADAM GROHS.
CO-FOUNDER/CEO

Particular.
the innovation agency of tomorrow, today

www.uniquelyparticular.com

**Comment by agrohs** [comment]
Thanks @notrab, I'm happy to share parts of our implementation offline if helpful, but may be best from a test case side to just try one of the custom implementations on npm (https://www.npmjs.com/package/node-fetch <https://github.com/github/fetch>) or the standard fetch polyfill (https://github.com/github/fetch <https://github.com/github/fetch>). There may even be test cases in one of those repos that would be helpful to grab for a basic nock or something?

File: src/config.js (hunk 2 above)
Comment (agrohs) [reply src/config.js]: Yes, I’m not partial as to what it’s called (actually had as something like custom_fetch earlier, but as it is used for the auto call was playing with auth_fetch or secure_fetch or credentials_fetch).

File: src/factories/request.js (hunk 1 above)
Comment (agrohs) [reply src/factories/request.js]: This is actually related to what I believe may be a defect in the SDK, I can remove the commented out block to leave as is but love to discuss as I think the current implementation may have an issue.

File: src/config.js (hunk 2 above)
Comment (agrohs) [reply src/config.js]: Updated name and committed as custom_fetch

File: src/factories/request.js (hunk 1 above)
Comment (agrohs) [reply src/factories/request.js]: Committed removal of comment about potential issue around credential expiration.

**Comment by agrohs** [comment]
Just checking in on this PR @notrab, @gje4, @ynnoj?

**Comment by agrohs** [comment]
@notrab, committed the additional changes/tests we discussed into the fork

**Comment by agrohs** [comment]
added missing catch clause on authenticate promise (need to double check, but currently have this deleting `moltinCredentials` from storage - could also just leave storage as is and do nothing in the catch clause, but seems if there is an auth error, that we would want to remove the previous stored credentials.

also, added the missing test that checks to ensure that if an invalid custom_fetch is passed in that it is rejected as a TypeError!

cc: @gje4, @notrab

**Comment by notrab** [comment]
Looks good to me!

File: src/factories/request.js [hunk 3]
```diff
@@ -65,14 +66,18 @@ class RequestFactory {
...
+      .then(response => {
+        const credentials = new Credentials(
+          config.client_id,
+          response.access_token,
+          response.expires
+        )
+        storage.set('moltinCredentials', JSON.stringify(credentials))
+      })
+      .catch(() => {
+        storage.delete('moltinCredentials')
```
Comment (notrab) [inline src/factories/request.js]: @agrohs Could you remove this? We can then apply this in a separate PR.

Comment (agrohs) [reply src/factories/request.js]: @notrab, shouldn't remove the catch itself as it is throwing an `UnhandledPromiseRejection` when a bad fetch is passed in (as tested in the unit test), but can remove the `storage.delete('moltinCredentials')` inside and you guys can figure out whether that is a desired fix/feature to add in on the moltin end at some point. Will update commit now w/ the empty catch block now per our discussion.

**Pull request merged**